adapters/beachfront/beachfront.go

```diff
@@ -0,0 +1,495 @@
+	errs := make([]error, 0, len(request.Imp))
+
+	var beachfrontRequests BeachfrontRequests
+	var reqJSON []byte
+	var uri string
+
+	uri = getEndpoint(request)
+
+	beachfrontRequests, err := preprocess(request, uri)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	if uri == VideoEndpoint {
+		reqJSON, err = json.Marshal(beachfrontRequests.Video)
+		uri = uri + beachfrontRequests.Video.AppId + VideoEndpointSuffix
+	} else {
+		/*
+			We will get here if request contains no Video imps, though it might have
+			Audio or Native imps as well as banner.
+		*/
+		reqJSON, err = json.Marshal(beachfrontRequests.Banner)
+	}
+
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if uri == BannerEndpoint {
+		if request.User != nil {
+			headers.Add("Cookie", "UserID="+request.User.ID+"; __io_cid="+request.User.BuyerUID)
+		}
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     uri,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+/*
+We have received a prebid request. It needs to be converted to a beachfront request. This is complicated
+by the fact that we have different servers for video/display and they have different contracts.
+*/
+func preprocess(req *openrtb.BidRequest, uri string) (BeachfrontRequests, error) {
+	var beachfrontReqs BeachfrontRequests
+	var err error
+
+	if uri == BannerEndpoint {
+		beachfrontReqs.Banner, err = getBannerRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	} else {
+		// If there were any Video imps in the request, we have skipped to here.
+		beachfrontReqs.Video, err = getVideoRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	}
+
+	return beachfrontReqs, err
+}
+
+func getBannerRequest(req *openrtb.BidRequest) (BeachfrontBannerRequest, error) {
+	var bannerImpsIndex int = 0
+	var beachfrontReq BeachfrontBannerRequest = NewBeachfrontBannerRequest()
+	var err error
+
+	/*
+	 step through the prebid request "imp" and inject into the beachfront request. If we got to here,
+	 then we have already stepped through the requested imps and verified that none are Video, so no
+	 reason to check that here, but there could be Audio or Native (or maybe they are filtered out before
+	 I get here based on the capabilities in bidder-info/beachfront.yaml? Regardless, I'll leave
+	 place holders here) .
+	*/
+
+	for _, imp := range req.Imp {
+		if imp.Audio != nil {
+			return beachfrontReq, &adapters.BadInputError{
+				Message: fmt.Sprintf("Beachfront doesn't support audio Imps. Ignoring Imp ID=%s", imp.ID),
+			}
+		} else if imp.Native != nil {
+			return beachfrontReq, &adapters.BadInputError{
+				Message: fmt.Sprintf("Beachfront doesn't support native Imps. Ignoring Imp ID=%s", imp.ID),
+			}
```

Comment:
Until #555 is fixed, these returns will mean you will not bid on any banner imps on a multi-imp request even if only one of the imps is an unsupported format. Not fatal, but I am not sure that is what you want.

Reply:
That is not what I want. Please hold off until my next push.